README.md
# ACSE (Advanced Compiler System for Education)

ACSE is a complete toolchain consisting of a compiler (named ACSE), an
assembler, and a simulator (named MACE). It provides a simple but reasonably
accurate sandbox for learning how compilers work.

## How to use ACSE

ACSE was tested on the following operating systems:

- **Linux** (any recent 32 bit or 64 bit distribution should work)
- **macOS** (any recent version should work)
- **Windows** (both 32 bit or 64 bit) when built with
  [MinGW](http://www.mingw.org) under [MSYS2](https://www.msys2.org), or inside
  **Windows Services for Linux** (WSL).

If you are using **Linux** or **macOS**, ACSE requires the following programs
to be installed:

- a **C compiler** (for example *GCC* or *clang*)
- **GNU bison**
- **GNU flex**

If you use **Windows**, first you must install either the
[MSYS2](https://www.msys2.org) environment or **Windows Services for Linux**
(WSL). Both MSYS2 and Windows Services for Linux (WSL) provide a Linux-like
environment inside of Windows.

Once you have installed either MSYS2 or WSL, you can use the following
instructions just as if you were using Linux or macOS.

### Building ACSE

To build the ACSE compiler toolchain, open a terminal and type:

      make

The built executables will be located in the `bin` directory.

### Testing ACSE

To compile some examples (located in the directory `tests`) type:

      make tests

### Using ACSE

In order to use the compiler/assembler/simulator, first you have
to export the directory `./bin` in your current `PATH` as follows:

      export PATH=`pwd`/bin:$PATH

You can compile and run new Lance programs in this way (suppose you
have saved a Lance program in `myprog.src`):

      acse myprog.src myprog.asm
      asm myprog.asm myprog.o
      mace myprog.o

You can invoke `acse`, `asm` and `mace` without setting `PATH` if you wish. In
that case you'll need to specify the path to where they are located.

For example, if the current directory is still the directory where you have
invoked `make` to build ACSE, you'll use the following commands:

      ./bin/acse myprog.src myprog.asm
      ./bin/asm myprog.asm myprog.o
      ./bin/mace myprog.o

Alternatively, you can add a test to the `tests` directory by following these
steps:

1. Create a new directory inside `tests`. You can choose whatever directory
   name you wish, as long as it is not `test`.
2. Move the source code files to be compiled inside the directory you have
   created. The extension of these files **must** be `.src`.
3. Run the command `make tests` to compile all tests, included the one you have
   just added.
   
The `make tests` command only runs the ACSE compiler and the assembler, you
will have to invoke the MACE simulator manually.


ReleaseNotes.md
# Release Notes

## ACSE 1.2.3

Released on 2021-12-15

- Fixed a bug in ACSE where multiple assignments of the same label did not
  generate an error.
- Fixed a bug in the assembler where multiple definitions of the same label did
  not generate an error.
- In the assembler, return a non-zero code from `main()` in case of error.
- Minor corrections to the comments in ACSE.
- Minor cleanup of the code related to labels.

## ACSE 1.2.2

Released on 2021-12-13

- Made shifts with amount greater than 31 and less than zero deterministic in
  both ACSE and MACE.
- Minor fixes to the makefiles.

## ACSE 1.2.1

Released on 2021-09-13

### New Features

- Now almost every header file in acse has a comment describing its purpose.
- Strings passed to `yyerror()` are now displayed rather than discarded.
- Improved formatting of the acse logs and the assembly printout.
- Changed file extension of the acse logs to `.log` from `.cfg` and `.out`.
- The name of the acse log files is now determined from the name of the input
  source code file to be compiled.
- Show a warning when the source code contains a shift amount that is too large.

### Bug Fixes

- Fixed documentation of the `gen_bhi_instruction()`, `gen_bls_instruction()`,
  `gen_bcs_instruction()`, and `gen_bgt_instruction()` functions.
- Acse no longer crashes when the source code contains a division by zero in a
  constant expression.
- Fixed semantics of the `.SPACE` assembler directive to take a number of words
  rather than a number of bytes.
- Removed `SEGMENTED` mode from MACE, as it did not properly load the data
  segment of executables.
- Other minor fixes to documentation and comments.

## ACSE 1.2.0

Released on 2020-09-14

### New Features

**Changes to acse:**

- The logical `NOT` operator (!) now supports expression arguments.
- Added clang-format and editorconfig files for automatic configuration of
  indentation and linting in several editors.
- The debug dump files and the compiled assembly now contains comments that map
  the instructions to the line in the source code that produced them.
- Labels corresponding to variables are now named like the variable in dump
  files and the compiled assembly.
- `collections`: added new `addBefore()` and `addAfter()` functions for adding
  an element before or after a given element in a list in O(1) time.
- `collections`: added new macros (`INTDATA`, `LINTDATA` and `SET_INTDATA`) for
  simplifying the use of the list data pointer as an integer.
- `axe_engine`: added new functions `newNamedLabel()` and
  `assignNewNamedLabel()` to create labels with a personalized name. 
- It is now possible to use immediates of any size (not restricted to 16 bits
  anymore) when generating instructions with immediate operands (`ADDI`, `SUBI`,
  `MULI`...)
- Add a new debug output file named `frontend.out` containing the state of the
  IR just after parsing, before any transformation takes place.
- acse now prints the version number on startup.
- Added support for computing def/use of the PSW (flags) register, as a
  debugging aid (the PSW register now appears in `dataflow.cfg`)
- Debug files are now flushed more frequently to ensure complete logs in case
  of crashes.
- Documentation improvements (fixed typos and grammar mistakes, clarified
  descriptions of some functions)
- Other API additions for internal use.

**Changes to mace:**

- Implemented `JSR` and `RET` instructions.
- Add a new instruction (`XPSW`) which exchanges the value of the PSW register
  with the value of a general-purpose register.
- When compiling mace with `DEBUG` enabled, fflush `stderr` at every instruction
  executed to ensure complete logs in case of crashes.
- Added a verification tool to use for regression testing. It can be executed
  by running `make verify` inside the mace directory.
- All comments written in Italian were replaced by comments written in English.
- Minor improvements to the debug features.

**Changes to asm:**

- Comments are now allowed anywhere in assembly language sources (previously
  they were allowed only after an instruction)
- Added support for hexadecimal immediates in the form `0x[0-9A-Za-z]+`
- Added support for inserting multiple data words within the same `.WORD`
  directive (syntax: `.WORD <word_1> <word_2> ... <word_n>`)

**Changes to tests:**

- Renamed syntax error tests to more descriptive names.

**Other changes:**

- Code formatting improvements across the entire project.
- Improvements to the makefiles:
  - All makefiles can now use the standard configuration variables `CFLAGS`,
    `YFLAGS` (bison), `LFLAGS` (flex) and `LDFLAGS`.
  - The test makefile now rebuilds the tests when Acse has changed.
  - Changed the test makefile to be whitespace-friendly.
  - Other improvements

### Bug Fixes

**Changes to acse:**

- `handle_bin_numeric_op` now supports all binary operations used in `Acse.y`.
- Fixed `gen_move_immediate` and `gen_load_immediate` to properly handle
  positive numbers larger than 32767 with bit 15 set.
- Fix a bug where an initialized variable was not properly loaded if it was
  used and defined exactly once by the same instruction in the first basic
  block.
- Fix a bug where invalid tokens after a valid statement caused parsing to stop
  mid-program without an error.
- Fix a bug where line numbers were incorrectly counted on lines with C++ style
  comments.
- Fix a bug in register allocation materialization where spill store
  instructions were placed one instruction too early in basic blocks without an
  explicit terminator instruction.
- Fixed a bug where the register allocator would unexpectedly run out of spill
  registers because the liveness of registers allocated to RD operands was not
  tracked correctly.
- Handle `RET` instructions like `HALT` instructions when performing analysis
  passes.
- Properly generate a `NEG` instruction instead of a `SUB` instruction in
  `gen_neg_instruction`.
- Properly return `SY_LOCATION_UNSPECIFIED` from `getLocation()` even when

mace/decode.c
/*
 * Giovanni Agosta, Andrea Di Biagio
 * Politecnico di Milano, 2007
 *
 * decode.c
 * Formal Languages & Compilers Machine, 2007/2008
 *
 */
#include <stdlib.h>
#include "decode.h"

/* Code Names for debug printing  */
char *names[4][16] = {
      {"ADD", "SUB", "ANDL", "ORL", "EORL", "ANDB", "ORB", "EORB", "MUL", "DIV",
            "SHL", "SHR", "ROTL", "ROTR", "NEG", "SPCL"},
      {"ADDI", "SUBI", "ANDLI", "ORLI", "EORLI", "ANDBI", "ORBI", "EORBI",
            "MULI", "DIVI", "SHLI", "SHRI", "ROTLI", "ROTRI", "NOTL", "NOTB"},
      {"NOP", "MOVA", "JSR", "RET", "LOAD", "STORE", "HALT", "SEQ", "SGE",
            "SGT", "SLE", "SLT", "SNE", "READ", "WRITE", "XPSW"},
      {"BT", "BF", "BHI", "BLS", "BCC", "BCS", "BNE", "BEQ", "BVC", "BVS",
            "BPL", "BMI", "BGE", "BLT", "BGT", "BLE"}};

decoded_instr *decode(int the_instr)
{
   decoded_instr *instr = (decoded_instr *)malloc(sizeof(decoded_instr));
   /* decoding */
   instr->format = getbits(the_instr, 31, 30); /* Instruction format (2 bits) */
   instr->opcode = getbits(the_instr, 29, 26); /* Opcode (4 bits) */
   instr->dest = getbits(the_instr, 25, 21);   /* Destination register */
   instr->src1 = getbits(the_instr, 20, 16);   /* Source register 1 */
   instr->src2 = getbits(the_instr, 15, 11);   /* Source register 2 */
   instr->imm = getbits(the_instr, 15, 0);     /* 16 bits immediate value */
   instr->addr = getbits(the_instr, 19, 0);    /* 20 bits immediate address */
   instr->func = getbits(the_instr, 10, 0);    /* 11 extra bits, used to encode
                                                  extra info */
   normalizeValues(instr);
   return instr;
}

int func_carry(decoded_instr *instr)
{
   return instr->func & 1;
}

int func_is_unsigned(decoded_instr *instr)
{
   return instr->func & 2;
}

int func_indirect_dest(decoded_instr *instr)
{
   return instr->func & 4;
}

int func_indirect_src2(decoded_instr *instr)
{
   return instr->func & 8;
}

/* Debug printf: print out the current instruction */
void print(FILE *file, decoded_instr *instr)
{
   /* default value for file is stderr */
   if (file == NULL)
      file = stderr;

   /* precondition: instr should be different from NULL */
   if (instr == NULL) {
      fprintf(file, "[NULL POINTER] \n");
      return;
   }

   switch (instr->format) {
      case FORMAT_TER:
         fprintf(file, "%s ", names[instr->format][instr->opcode]);
         if (func_indirect_dest(instr))
            fprintf(file, "(R%d) ", instr->dest);
         else
            fprintf(file, "R%d ", instr->dest);
         fprintf(file, "R%d ", instr->src1);
         if (func_indirect_src2(instr))
            fprintf(file, "(R%d) \n", instr->src2);
         else
            fprintf(file, "R%d \n", instr->src2);
         break;
      case FORMAT_BIN:
         fprintf(file, "%s R%d, R%d, #%d \n",
               names[instr->format][instr->opcode], instr->dest, instr->src1,
               instr->imm);
         break;
      case FORMAT_UNR:
         if (instr->opcode == 0 || instr->opcode == 6)
            fprintf(file, "%s \n", names[instr->format][instr->opcode]);
         else if (instr->opcode == 13 || instr->opcode == 14)
            fprintf(file, "%s R%d \n", names[instr->format][instr->opcode],
                  instr->dest);
         else
            fprintf(file, "%s R%d, #%d\n", names[instr->format][instr->opcode],
                  instr->dest, instr->addr);
         break;
      case FORMAT_CC:
         fprintf(file, "%s $ %c %d\n", names[instr->format][instr->opcode],
               instr->addr < 0 ? '-' : '+', abs(instr->addr));
         break;
      default: fprintf(file, "[UNKNOWN INSTRUCTION FORMAT] \n"); break;
   }
}

void normalizeValues(decoded_instr *instr)
{
   /* sign extension of a 16 bit value over 32 bit */
   instr->imm = (instr->imm & 0xffff);
   if (instr->imm & 0x8000)
      instr->imm |= 0xffff0000;

   /* sign extension of a 20 bit value over 32 bit */
   instr->addr = instr->addr & 0x0fffff;
   if (instr->addr & 0x080000)
      instr->addr |= 0xfff00000;
}
